firewood/src/storage/linear/mod.rs

```diff
@@ -68,47 +70,60 @@ mod filebacked;
 mod historical;
 mod layered;
 mod proposed;
-#[cfg(test)]
-mod tests;
 
 #[derive(Debug)]
-pub(super) struct LinearStore<S: ReadLinearStore> {
-    state: S,
-}
-
-/// All linearstores support reads
-pub(super) trait ReadLinearStore: Send + Sync + Debug {
-    fn stream_from(&self, addr: u64) -> Result<Box<dyn Read + '_>, Error>;
-    fn size(&self) -> Result<u64, Error>;
+pub(super) enum ImmutableLinearStore {
+    Historical(Historical),
+    Proposed(Arc<Proposed<Immutable>>),
+    FileBacked(FileBacked),
+    Invalid, // TODO do we need this variant?
 }
 
-impl ReadLinearStore for Arc<dyn ReadLinearStore> {
-    fn stream_from(&self, addr: u64) -> Result<Box<dyn Read + '_>, Error> {
-        self.deref().stream_from(addr)
+impl ImmutableLinearStore {
+    pub fn stream_from(&self, addr: u64) -> Result<Box<dyn Read + '_>, Error> {
+        match self {
+            ImmutableLinearStore::Historical(historical) => historical.stream_from(addr),
+            ImmutableLinearStore::Proposed(proposed) => proposed.stream_from(addr),
+            ImmutableLinearStore::FileBacked(filebacked) => filebacked.stream_from(addr),
+            ImmutableLinearStore::Invalid => Err(std::io::ErrorKind::InvalidData.into()),
+        }
     }
 
-    fn size(&self) -> Result<u64, Error> {
-        self.deref().size()
+    pub fn size(&self) -> Result<u64, Error> {
+        match self {
+            ImmutableLinearStore::Historical(historical) => historical.size(),
+            ImmutableLinearStore::Proposed(proposed) => proposed.size(),
+            ImmutableLinearStore::FileBacked(filebacked) => filebacked.size(),
+            ImmutableLinearStore::Invalid => Err(std::io::ErrorKind::InvalidData.into()),
+        }
```

Comment:
🤔, if you're only methods are delegating, why not just use a trait object? Then you don't need the match statement. You can implement `ReadLinearStore` for `Arc<T>` where `T: ReadLinearStore`.